src/python/pants/backend/jvm/tasks/check_banned_deps.py

```diff
@@ -0,0 +1,166 @@
+    self._constraints = direct + transitive
+
+  def _compute_fingerprint(self):
+    return combine_hashes([c.fingerprint() for c in self._constraints])
+
+  def check_all(self, target, task_context):
+    # TODO Eventually, the next three lines should be moved out of here,
+    #      and the `classes_relevant_to_target` parameter removed from `check_target`.
+    #      Ideally we would have the same interface for all the subclasses of DependencyConstraint,
+    #      which I think should be check_target(target, task_context)
+    classpath = task_context.products.get_data("runtime_classpath")
+    analyzer = JvmDependencyAnalyzer(get_buildroot(), classpath)
+    classes_relevant_to_target = analyzer.classes_for_targets(DependencyConstraints.relevant_targets(target))
+    for constraint in self._constraints:
+      constraint.check_target(target, task_context, classes_relevant_to_target)
+
+  @staticmethod
+  def relevant_targets(target):
+    """
+    Modify this method when the criteria changes
+    (i.e. the target itself should be included in the checks).
+    """
+    return target.dependencies
+
+
+class DependencyConstraint(object):
+  """Umbrella to hold the Constraint class hierarchy.
+
+  Has no functionality, but allows targets to use syntax like
+    `DependencyConstraint.JvmPackageConstraint("com.google")`
+  """
+  class BannedDependencyException(Exception):
+    pass
+
+  class Constraint(PayloadField):
+    """Representation of a constraint on the target's dependencies."""
+
+    @abstractmethod
+    def check_target(self, target, task_context, classes_relevant_to_target):
+      pass
+
+  class JvmPackageConstraint(Constraint):
+
+    def __init__(self, name):
+      self.banned_package_name = name
+
+    def check_target(self, target, context, classes_relevant_to_target):
+      """
+      Check whether a given target complies with this constraint.
+      :param target: Target to check.
+      :param task_context: Context of the task, to extract from it any relevant information.
+      :param classes_relevant_to_target: Fully qualified names of the classes in the target and its dependencies.
+      :return:
+      """
+      banned_classes = [c for c in classes_relevant_to_target if c.startswith(self.banned_package_name)]
+      if banned_classes:
+        raise DependencyConstraint.BannedDependencyException(
+          'Target {} uses rule "{}" to ban classes ({})'.format(
+            target.target_base,
+            self.banned_package_name,
+            ", ".join(banned_classes)
+          ))
+
+    def _compute_fingerprint(self):
+      return stable_json_hash(self.banned_package_name)
+
+  class Tag(Constraint):
+    """Representation of the ban of a tag"""
+
+    def __init__(self, banned_tag_name):
+      self.banned_tag_name = banned_tag_name
+
+    def _has_bad_tag(self, target):
+      return self.banned_tag_name in target.tags
+
+    def check_target(self, target, task_context, classes_relevant_to_target):
```

Comment:
How about, instead of including `classes_relevant_to_target` as a default parameter, you extract out the common bits of all of the check_target methods into the super class and pass in a predicate function?